compiler/rustc_mir_transform/src/check_unsafety.rs

```diff
@@ -109,6 +109,11 @@ impl<'tcx> Visitor<'tcx> for UnsafetyChecker<'_, 'tcx> {
                 // safe (at least as emitted during MIR construction)
             }
 
+            StatementKind::Assume(..) => self.require_unsafe(
+                UnsafetyViolationKind::General,
+                UnsafetyViolationDetails::CallToUnsafeFunction,
+            ),
```

Comment:
If the next step is to add assume statements during MIR building this should be simply allowed with a comment explaining that it is only built for safe constructs. 

Right now this code is unreachable since the unsafety checks precede intrinsics lowering. The order of those passes is quite important given that unsafe `unreachable` intrinsic is lowered to unreachable terminator which is considered safe.